hsigmoid/READ_EN.md

```diff
@@ -0,0 +1,174 @@
+        hidden_layer, param_attr=paddle.attr.Param(name="sigmoid_w")),
+    act=paddle.activation.Sigmoid(),
+    bias_attr=paddle.attr.Param(name="sigmoid_b"))
+```
+
+In above code slice, `paddle.layer.mixed` must use `paddle.layer.×_projection` of PaddlePaddle as input.`paddle.layer.mixed` summate multiple computing results of `projection`(input can be more than one) as output.`paddle.layer.trans_full_matrix_projection` will transpose parameter $W$ when computing matrix multiplication.
+
+
+## Training stage
+Training is easy, directly run ``` python train.py ``` . Program will chech if imikolov data is included in user cache directory when first running, if not, download automatically. When running, it will print model training information every 100 iteration, mainly concerns training loss and testing loss, saves model for each pass.
+
+## Predicting stage
+In command line, run:
+
+```bash
+python infer.py \
+  --model_path "models/XX" \
+  --batch_size 1 \
+  --use_gpu false \
+  --trainer_count 1
+```
+parameters meaning:
+- `model_path`：assign path for well-trained model. Required.
+- `batch_size`：num of sample items for one parallel predicting.Optional, default `1`
+- `use_gpu`：if use GPU for prediction or not. Optional, default `False`
+- `trainer_count` : num of thread for prediction. Optional, default `1`**
+
+
+In training stage, we can get code path by multiple binary classification propability, iterate path to get final class, reason is:
+
+```python
+def decode_res(infer_res, dict_size):
+    """
+    Inferring probabilities are orginized as a complete binary tree.
+    The actual labels are leaves (indices are counted from class number).
+    This function travels paths decoded from inferring results.
+    If the probability >0.5 then go to right child, otherwise go to left child.
+
+    param infer_res: inferring result
+    param dict_size: class number
+    return predict_lbls: actual class
+    """
+    predict_lbls = []
+    infer_res = infer_res > 0.5
+    for i, probs in enumerate(infer_res):
+        idx = 0
+        result = 1
+        while idx < len(probs):
+            result <<= 1
+            if probs[idx]:
+                result |= 1
+            if probs[idx]:
+                idx = idx * 2 + 2  # right child
+            else:
+                idx = idx * 2 + 1  # left child
+
+        predict_lbl = result - dict_size
+        predict_lbls.append(predict_lbl)
+    return predict_lbls
+```
+Format of input data for predicting is same with training, like `have a dream that one`, program will produce a sequence of propability based on `have a dream that`, then decode propability to generate predicting word, `one` as real word, is easy to evaluate. Input of decoding funtion is prediction propability of one batch sample and vocabulary size, inner loop is to decode each item based on its output propability, decoding method follows left 0 right 1 rule, iterates all paths until arrives at leaf node.
+
+
+## Reference
```

Comment:
Data Preparation

PTB Data

This article uses the Penn Treebank (PTB) data set (Pre-Tomas Mikolov version), which contains three files: train, valid, and test. We use 'train' as training data and 'valid' as test data. This article trains the 5-gram model, which uses the first 4 words of each piece of data to predict the 5th word. PaddlePaddle provides a python package, paddle.dataset.imikolov, corresponding to the PTB dataset, which automatically performs data download and preprocessing. The preprocessing operations will add a start symbol <s> and an end symbol <e> before and after each sentence in the data set and, depending on the window size (in this article, the size is 5), swipe the window to the right from the beginning to the end to generate a piece of data. For instance, "I have a dream that one day" can generate these pieces of data: <s> I have a dream, I have a dream that, have a dream that one, a dream that one day, dream that one day <e>. PaddlePaddle will output the ID information of the words.

Custom Data
Users can use their own data set to train the model. The most important feature of a custom data set is to implement the reader interface for data processing. The reader needs to generate an iterator that is responsible for parsing each row of data in the file and returning a python list. For example, suppose that 1, 2, 3, and 4 are the IDs of the first to fourth word in the dictionary. PaddlePaddle will convert the list of IDs into paddle.data_type.inter_value type as the data layer input. A package example is as the following:


Network Structure

This article explains how to obtain the word vector by training the N-gram language model. We use the first four words to predict the current word. The network input is the ID of the word in the dictionary. The query word vector table is used to obtain the word vector. We stitch together the four word vectors and access a fully connected hidden layer. Finally, we get to the Hsigmoid layer. The detailed network structure is shown in Figure 2:


Figure 2. Network configuration structure

code is shown as follows:


It should be noted that in PaddlePaddle, the hsigmoid layer stores the learnable parameters as a matrix of [number of classes - 1 × hidden layer vector width]. When forecasting, you need to replace the hsigmoid layer with a full-join operation and fix it with sigmoid. Prediction has the output as a matrix with width of [batch_size x number of categories - 1] (degenerate into a vector when batch_size = 1). Each dimension of a matrix row vector computes the probability of a right child node whose input vector belongs to an internal node. When the full-joined operation loads the parameter matrix learned by the hsigmoid layer, the parameter matrix needs to be transposed once. The code fragment is as follows:


The paddle.layer.mixed in the above code snippet must be entered as paddle.layer.×_projection in PaddlePaddle. Paddle.layer.mixed sums the results of multiple projection (there can be multiple inputs) as output. Parameter W is transposed by paddle.layer.trans_full_matrix_projection when doing matrix multiplication.

Training Phase

The training is relatively simple; we can run python train.py directly. The first time the program runs, it will check if the user's cache folder contains the imikolov data set. If the data set is not present, it will be downloaded automatically. During the running process, model training information will be printed every 100 iterations, which mainly includes training loss and test loss. Each pass will save the model once.

Prediction Stage

Run the following command line:


The parameters are as follows:
 - model_path: Specifies the path where the trained model is located. This parameter is required.
 - batch_size: Predicts the number of samples in parallel at a time. Optional with default value 1.
- use_gpu: Whether to use GPU for prediction. Optional, the default value is False. - trainer_count : The number of threads predicted to use. Optional, default is 1. Note: The number of threads used for prediction must be greater than the number of samples to be predicted in parallel.
The prediction stage obtains the encoding path based on multiple binary classification probabilities and traverses the path to obtain the final prediction category. The logic is as follows:

The format of input data during the prediction phase is the same as it is in the training phase. For instance, for the phrase "have a dream that one", the program will generate a set of probabilities based on the phrase. By inspecting the probabilities, the program predict the output words. The input of the decoding function is the prediction probability of a batch sample and the size of the vocabulary. The loop inside the decoding function decodes the output probability of each sample. The decoding method follows that if meeting a zero, we go to the left child node and if meeting a one, we go to the right child. We continue to traverse the path until reaching a leaf node. .

References